### PR TITLE
PP-357/dapp-update

### PR DESCRIPTION
**Pull request by franciscotobar**
## What

- changes to support dApp refactoring

## Why

- required by dApp refactor

## Refs

- [update rif-relay-sample-dapp](https://rsklabs.atlassian.net/browse/PP-357)

### CODE REVIEW

File: src/RelayClient.ts [hunk 1]
```diff
@@ -38,57 +44,83 @@ import { isDeployRequest } from './common/relayRequest.utils';
...
+      throw new Error('Field `from` is not defined in request body.');
+    }
+
+    if (!tokenContract) {
+      throw new Error('Field `tokenContract` is not defined in request body.');
+    }
+
+    if (!isDeployment) {
+      if (!relayData?.callForwarder) {
+        throw new Error(MISSING_CALL_FORWARDER);
```
Comment (antomor) [inline src/RelayClient.ts:98]: Why do we store only this error message in the constant? I'd try to be consistent and do the same with the others.

File: src/RelayClient.ts [hunk 2]
```diff
@@ -153,40 +168,35 @@ class RelayClient extends EnvelopingEventEmitter {
...
       callVerifier,
       feesReceiver: constants.AddressZero, // returns zero address and is to be completed when attempting to relay the transaction
       gasPrice,
     };
 
     const secondsNow = Math.round(Date.now() / 1000);
-    const validUntilTime = (
-      secondsNow + this._envelopingConfig.requestValidSeconds
-    );
+    const validUntilTime = request.validUntilTime ??
```
Comment (antomor) [inline src/RelayClient.ts:198]: @franciscotobar thank you for this fix 👍🏽

File: package.json [hunk 3]
```diff
@@ -54,11 +54,9 @@
     "package.json": "npx sort-package-json"
   },
   "dependencies": {
-    "@metamask/eth-sig-util": "^5.0.0",
     "@rsksmart/rif-relay-contracts": "github:rsksmart/rif-relay-contracts#feature/PP-278/PP-329/upgrade_contracts",
-    "axios": "~1.1.2",
+    "axios": "~0.27.2",
```
Comment (antomor) [inline package.json:58]: Should we add a task in the backlog to update the axios version?

File: src/RelayClient.ts [hunk 4]
```diff
@@ -219,25 +229,25 @@ class RelayClient extends EnvelopingEventEmitter {
...
-    const accountManager = new AccountManager(this._provider, chainId);
+      request: { relayHub, tokenGas, nonce, gas, value },
+      relayData: { gasPrice },
+    } = envelopingRequest;
+
+    const provider = getProvider();
+    const accountManager = new AccountManager();
     const relayMaxNonce =
-      (await this._provider.getTransactionCount(callForwarder)) +
+      (await provider.getTransactionCount(relayWorkerAddress)) +
```
Comment (antomor) [inline src/RelayClient.ts:250]: Another good fix, thank you 👍🏽

File: src/RelayClient.ts [hunk 5]
```diff
@@ -286,160 +316,133 @@ class RelayClient extends EnvelopingEventEmitter {
...
-          log.debug('Relay Client - Relayed done');
+    log.debug('Relay Client - Estimating transaction');
+    log.debug(
+      `Relay Client - Relay Hub:${envelopingRequestDetails.request.relayHub.toString()}`
+    );
 
-          return transaction;
-        }
-      }
+    const activeRelay = await selectNextRelay(this._httpClient);
```
Comment (antomor) [inline src/RelayClient.ts:408]: The FIXME needs to be applied here also

File: src/RelayClient.ts [hunk 6]
```diff
@@ -286,160 +316,133 @@ class RelayClient extends EnvelopingEventEmitter {
...
 
-      activeRelay = await selectNextRelay(this._httpClient);
-    }
+    const envelopingTx = await this._prepareHttpRequest(
+      activeRelay!.hubInfo,
+      envelopingRequestDetails
+    );
```
Comment (antomor) [inline src/RelayClient.ts:413]: Unless I'm missing something, it looks like this part is identical to what we have in the relayTransaction (L#365-376)

File: src/api/common/HttpWrapper.ts [hunk 7]
```diff
@@ -50,21 +50,23 @@ const interceptors = {
...
       }
-      log.debug(`Configuration: ${stringify(config)}`);
+      logger().debug(`Configuration: ${stringify(config)}`);
 
       return Promise.reject(error);
     },
   },
 };
 
+const logger = () => log.getLogger('HttpWrapper');
```
Comment (antomor) [inline src/api/common/HttpWrapper.ts:68]: I know it's the same, but I'd prefer to define the `logger` function at the top of the file.

File: src/utils.ts [hunk 8]
```diff
@@ -1,41 +1,30 @@
...
 
-  for (const managerData of preferredRelays ?? []) {
+  for (const preferredRelay of preferredRelays ?? []) {
     let hubInfo;
+    let managerData;
 
     try {
-      hubInfo = await httpClient.getChainInfo(managerData.url);
+      hubInfo = await httpClient.getChainInfo(preferredRelay);
+      const relayHub = RelayHub__factory.connect(hubInfo.relayHubAddress, getProvider());
```
Comment (antomor) [inline src/utils.ts:26]: Why don't we check any more for the `ready` property in `hubInfo`

File: src/utils.ts [hunk 9]
```diff
@@ -54,6 +43,123 @@ const selectNextRelay = async (
...
+    internalEstimationCorrection
+  );
+
+  return applyGasCorrectionFactor(
+    internalCallCost,
+    estimatedGasCorrectionFactor
+  );
+}
+
+const getSmartWalletAddress = async (
```
Comment (antomor) [inline src/utils.ts:122]: In the future, we should consider if we want to replace this mechanism with the one that doesn't use the factory to check if the result is the same and if we can avoid using the factory.

File: test/RelayClient.test.ts [hunk 10]
```diff
@@ -515,6 +492,23 @@ describe('RelayClient', function () {
...
+            },
+          } as unknown as UserDefinedEnvelopingRequest
+        );
+
+        await expect(call).to.be.eventually.rejected;
+        await expect(call).to.be.eventually.rejectedWith(
+          'Field `index` is not defined in deploy body.'
+        );
+      });
+
```
Comment (antomor) [inline test/RelayClient.test.ts:511]: I'd add a specific case if the index is 0

File: test/RelayClient.test.ts [hunk 11]
```diff
@@ -189,12 +177,12 @@ describe('RelayClient', function () {
...
             relayHub: expectedRelayHubAddress,
           },
-        }, FAKE_REQUEST_CONFIG);
+        });
 
         expect(actualRelayHubAddress).to.equal(expectedRelayHubAddress);
       });
 
-      it('should return tx request with fees receiver in request data from given hub info if no fees receiver given in request', async function () {
+      it.skip('should return tx request with fees receiver in request data from given hub info if no fees receiver given in request', async function () {
```
Comment (antomor) [inline test/RelayClient.test.ts:185]: How do we plan to fix the skipped test? Do we want to add a specific task for all of them?

**Review by antomor — CHANGES_REQUESTED** [review]
@franciscotobar Thank you for these changes. I left some comments to keep in mind that there are still open points there.

File: package.json (hunk 3 above)
Comment (franciscotobar) [reply package.json:58]: yes I think so, I tried but it was time consuming so I decided to follow the easy path.

File: src/RelayClient.ts (hunk 5 above)
Comment (franciscotobar) [reply src/RelayClient.ts:408]: updated

File: src/api/common/HttpWrapper.ts (hunk 7 above)
Comment (franciscotobar) [reply src/api/common/HttpWrapper.ts:68]: updated

File: src/utils.ts (hunk 8 above)
Comment (franciscotobar) [reply src/utils.ts:26]: we do in line 33, but I did some refactor.

File: test/RelayClient.test.ts (hunk 11 above)
Comment (franciscotobar) [reply test/RelayClient.test.ts:185]: now I forgot to discuss it with you, maybe after the daily we can check it.

File: src/RelayClient.ts (hunk 6 above)
Comment (franciscotobar) [reply src/RelayClient.ts:413]: now is the same, but before it was different.

File: src/RelayClient.ts (hunk 1 above)
Comment (franciscotobar) [reply src/RelayClient.ts:98]: updated

File: src/RelayClient.ts (hunk 5 above)
Comment (franciscotobar) [reply src/RelayClient.ts:408]: updated

File: src/RelayClient.ts (hunk 6 above)
Comment (franciscotobar) [reply src/RelayClient.ts:413]: updated

File: src/utils.ts (hunk 8 above)
Comment (franciscotobar) [reply src/utils.ts:26]: updated

File: src/utils.ts (hunk 9 above)
Comment (franciscotobar) [reply src/utils.ts:122]: updated

File: test/RelayClient.test.ts (hunk 11 above)
Comment (franciscotobar) [reply test/RelayClient.test.ts:185]: updated

File: test/RelayClient.test.ts [hunk 12]
```diff
@@ -865,70 +897,173 @@ describe('RelayClient', function () {
...
+
+      it('should build a transaction using a hub', async function () {
+        sandbox.stub(relayUtils, 'selectNextRelay').resolves(relayInfo);
+        relayClient._prepareHttpRequest = sandbox.stub().returns(FAKE_RELAY_TRANSACTION_REQUEST);
+        const enveloping = await relayClient._getHubEnvelopingTx(envelopingRequest);
+
+        expect(enveloping).to.be.deep.equal(expectedEnveloping);
+      });
+
+      it('should throw if there is no there is no hub available', async function () {
```
Comment (antomor) [inline test/RelayClient.test.ts:1052]:
```suggestion
      it('should throw if there is no hub available', async function () {
```

**Review by antomor — APPROVED** [review]

**Review by antomor — APPROVED** [review]
LGTM

**Pull request merged**